Enforce lightning invoice expiry

wilfredallyn opened the pull request:
Fixes https://github.com/fedimint/fedimint/issues/749

In previous PR (https://github.com/fedimint/fedimint/pull/727), we exposed the ability to set a custom invoice expiry but did not enforce it. This PR enforces the expiry by adding a new `LightningConsensusItem` for clock time

Todo:
- ~Need to fix integration test (https://github.com/fedimint/fedimint/commit/4eb650a8b47abc1b404bb7c97bf8b49ec786293b) for checking that the expiry is enforced~
- ~Added epochs to integration tests that were timing out (https://github.com/fedimint/fedimint/commit/614ca5163142252f1d8ee0a9a7da3c5b68515966), need to confirm that they still work as intended~

wilfredallyn commented on integrationtests/tests/tests.rs:
```diff
@@ -374,7 +376,7 @@ async fn drop_peers_who_dont_contribute_decryption_shares() -> Result<()> {
         user.client
             .generate_invoice(payment_amount, "".into(), rng(), None),
-        fed.await_consensus_epochs(1) // create offer
+        fed.await_consensus_epochs(2) // create offer
```
The number of epochs were increased due to tests timing out. It seems like adding the consensus item for clock time requires adding a consensus round, but this is my first time working on consensus code so not 100% sure.

elsirion replied:
Need to look into this in more detail, maybe @jkitman has some intuition.

jkitman replied:
It looks like after adding `LightningConsensusItem`, the LN module will continually produce epochs all the time.

However, we want to avoid this behavior to reduce the amount of unnecessary bandwidth/storage/compute that nodes use.

I suggest modifying `await_consensus_proposal` in the LN module to continue only when the time will generate a change in invoice expiry.

wilfredallyn replied:
Added logic in [`consensus_proposal`](https://github.com/fedimint/fedimint/pull/1074/commits/1ea20954bb029c09c36cf2febdcf621f064a5bac#diff-d777522b5e9ff491a0359a5506622730cabc67755d33cfab1a5ebf3cf1b0738dR309-R337) to only propose new clock time if an invoice would expire to avoid continually producing epochs

wilfredallyn commented on integrationtests/tests/tests.rs:
```diff
@@ -722,6 +726,34 @@ async fn set_lightning_invoice_expiry() -> Result<()> {
 }
 
+#[tokio::test(flavor = "multi_thread")]
+async fn client_cannot_pay_expired_invoice() -> Result<()> {
```
This test is a sketch of checking if the expiry is enforced, but it doesn't work yet. The invoice expiry it set to 1 second, and it sleeps for 2 seconds so the invoice should be expired when fund_outgoing_ln_contract is called

wilfredallyn replied:
Fixed test in https://github.com/fedimint/fedimint/pull/1074/commits/d1bf57e966283cd30cb3222fb307735dd9585c3e

elsirion commented on modules/fedimint-ln/src/lib.rs:
```diff
@@ -939,6 +1000,25 @@ impl LightningModule {
+        assert!(!proposals.is_empty());
+
+        proposals.sort();
+        let median_proposal = match proposals.len() % 2 {
+            0 => {
+                let earlier_time = proposals[proposals.len() / 2 - 1];
+                let later_time = proposals[proposals.len() / 2];
+                let time_diff = later_time.duration_since(earlier_time);
+                earlier_time
+                    + time_diff
+                        .expect("Error calculating median time")
+                        .div_f32(2.0)
+            }
+            _ => proposals[proposals.len() / 2],
+        };
```
That might be a bit over the top, I'd rather avoid the floating point math. As long as the chosen element is between the elements of two honest parties it's ok (e.g. with elements 1,2,3,4 and up to one malicious party both 2 and 3 are ok to choose). Just choose one.

wilfredallyn replied:
I simplified the clock time proposal in https://github.com/fedimint/fedimint/pull/1074/commits/24a5d5e2595e51a17b2d21a910cd0361e5834f08

elsirion commented on modules/fedimint-ln/src/lib.rs:
```diff
@@ -208,17 +209,46 @@ pub struct LightningGateway {
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize, Encodable, Decodable)]
+pub struct RoundConsensusItem {
+    pub clock_time: SystemTime,
```
On reflection we might want to make this a `u64` or some other internal timestamp type to avoid platform-dependent behavior of `SystemTime` (maybe that's not the case, but should check, different OSs treat time differently)

wilfredallyn replied:
I converted clock_time to `u64` in https://github.com/fedimint/fedimint/pull/1074/commits/39304eadf2560d4f3d161d333f1d5a9666bafd0f

elsirion reviewed (commented):
Sorry for the long wait, this looks pretty good already :)

jkitman commented on integrationtests/tests/tests.rs:
```diff
@@ -722,6 +726,34 @@ async fn set_lightning_invoice_expiry() -> Result<()> {
+    let invoice = lightning.invoice(sats(1000), 1.into());
+    // This does not work correctly. It should sleep 2 seconds so that the invoice is expired when fund_outgoing_ln_contract is called
+    thread::sleep(Duration::from_secs(2));
+    fed.run_consensus_epochs(1).await;
```
It will pass if you run two consensus epochs here, you need 1 for peers to send out the new time and another for that to be updated in consensus.  Perhaps you need to rearrange code in the LN module to do it in 1 epoch.

wilfredallyn replied:
Fixed hanging tests in https://github.com/fedimint/fedimint/pull/1074/commits/d1bf57e966283cd30cb3222fb307735dd9585c3e

jkitman reviewed (changes requested):
It looks like after adding `LightningConsensusItem`, the LN module will continually produce epochs all the time.

However, we want to avoid this behavior to reduce the amount of unnecessary bandwidth/storage/compute that nodes use.

I suggest modifying `await_consensus_proposal` in the LN module to continue only when the time will generate a change in invoice expiry (you can see how this is done in the wallet module).

Rather than have the client fetch the consensus time and validate the invoice client-side, we could have the federation validate and generate a consensus time on submission of the tx.

Overall looks good though! :smiley:

wilfredallyn commented on integrationtests/tests/tests.rs:
```diff
@@ -901,13 +933,14 @@ async fn lightning_gateway_cannot_claim_invalid_preimage() -> Result<()> {
         .iter()
         .filter(|item| match item {
             ConsensusItem::Module(mci) => mci.module_key() == MODULE_KEY_LN,
+            // ConsensusItem::Module(LightningConsensusItem::DecryptionShare(_)) => true,
```
Fixed test in https://github.com/fedimint/fedimint/pull/1074/commits/d1bf57e966283cd30cb3222fb307735dd9585c3e

wilfredallyn commented:
> I suggest modifying `await_consensus_proposal` in the LN module to continue only when the time will generate a change in invoice expiry (you can see how this is done in the wallet module).

Added logic in [consensus_proposal](https://github.com/fedimint/fedimint/pull/1074/commits/1ea20954bb029c09c36cf2febdcf621f064a5bac#diff-d777522b5e9ff491a0359a5506622730cabc67755d33cfab1a5ebf3cf1b0738dR309-R337) to only propose new clock time if an invoice would expire to avoid continually producing epochs. I put the logic in `consensus_proposal` instead of `await_consensus_proposal` to based on the logic changes I saw in the wallet .
 
> Rather than have the client fetch the consensus time and validate the invoice client-side, we could have the federation validate and generate a consensus time on submission of the tx.

I moved the validation logic from client-side to federation in https://github.com/fedimint/fedimint/pull/1074/commits/1ea20954bb029c09c36cf2febdcf621f064a5bac

wilfredallyn commented on modules/fedimint-ln-server/tests/ln_contracts.rs:
```diff
@@ -31,25 +35,22 @@ async fn test_outgoing() {
     let preimage = Preimage([42u8; 32]);
     let hash = secp256k1::hashes::sha256::Hash::hash(&preimage.0);
 
-    let invoice: lightning_invoice::Invoice =
```
Using an old invoice will cause an expired invoice error in [apply_output](https://github.com/fedimint/fedimint/pull/1074/commits/1ea20954bb029c09c36cf2febdcf621f064a5bac#diff-d777522b5e9ff491a0359a5506622730cabc67755d33cfab1a5ebf3cf1b0738dR506-R514)

I fixed it by creating an invoice with InvoiceBuilder which will have the default expiry based on the current time

elsirion commented on modules/fedimint-ln-server/src/lib.rs:
```diff
@@ -296,7 +310,32 @@ impl ServerModule for Lightning {
+            panic!("No proposals were submitted this round");
+        }
+
+        let clock_time_proposals = round_consensus
+            .iter()
+            .map(|(_, rc)| rc.clock_time)
+            .collect();
```
Before throwing away the peer id, do we make sure that each peer only contributed one item? Otherwise malicious ones could take over. Needs to be fixed in wallet too.

elsirion commented on modules/fedimint-ln-server/src/lib.rs:
```diff
@@ -859,6 +898,25 @@ impl Lightning {
     }
+
+    async fn process_clock_time_proposals(&self, mut proposals: Vec<SystemTime>) -> SystemTime {
+        assert!(!proposals.is_empty());
```
nit: Would be nice if all asserts had comments

elsirion commented on modules/fedimint-ln-server/src/lib.rs:
```diff
@@ -859,6 +898,25 @@ impl Lightning {
+                earlier_time
+                    + time_diff
+                        .expect("Error calculating median time")
+                        .div_f32(2.0)
```
I'd rather avoid floating point operations since they tend to slightly differ on some architectures (mostly due to CPU bugs afaik).

jkitman commented on integrationtests/tests/tests.rs:
```diff
@@ -788,6 +789,42 @@ async fn set_lightning_invoice_expiry() -> Result<()> {
+            .await_outgoing_contract_acceptance(outpoint)
+            .await;
+
+        if let Err(mint_client::ClientError::OutputOutcome(OutputOutcomeError::Rejected(reason))) =
```
nit: can you use `assert_matches`?

jkitman commented on integrationtests/tests/tests.rs:
```diff
@@ -954,7 +991,11 @@ async fn lightning_gateway_cannot_claim_invalid_preimage() -> Result<()> {
         assert!(response.is_err());
 
         fed.run_empty_epochs(1).await; // if valid would create contract to mint notes
-        assert_eq!(fed.find_module_item(fed.ln_id).await, None);
```
How come this changed?

jkitman commented on modules/fedimint-ln-server/src/lib.rs:
```diff
@@ -268,35 +278,96 @@ impl ServerModule for Lightning {
+        // We force new epochs only if invoice expires, or we have decryption share
+        // (more than just round_ci item)
+        if 1 < items.len()
+            || (outgoing_contracts
```
Looks better!  Could we just `.filter(|c| c.invoice.is_expired())` on `outgoing_contracts` and then check if `len() > 0`?

elsirion commented on modules/fedimint-ln-server/src/lib.rs:
```diff
@@ -799,6 +799,12 @@ impl ServerModule for Lightning {
+            api_endpoint! {
+                "/clock_time",
+                async |module: &LightningModule, dbtx, _params: ()| -> SystemTime {
+                   Ok(module.consensus_clock_time(&mut dbtx).await.unwrap_or(UNIX_EPOCH))
```
I'd rather return an `Option` here or have the `UNIX_EPOCH` be the default in all cases (i.e. happens in `consensus_clock_time`).

elsirion commented on client/client-lib/src/lib.rs:
```diff
@@ -833,6 +833,11 @@ impl Client<UserClientConfig> {
+
+        if invoice.would_expire(consensus_clock_time.duration_since(UNIX_EPOCH).unwrap()) {
+            return Err(ClientError::LnClientError(LnClientError::ExpiredInvoice));
+        }
```
For outgoing LN payments the recipient will tell the gateway that the invoice expired, which triggers a refund. We don't need to pre-check that necessarily. On the incoming side where we are the recipient we should do that on the other hand.

elsirion commented on client/client-lib/src/api.rs:
```diff
@@ -27,7 +27,7 @@ pub trait LnFederationApi {
         payment_hash: Sha256Hash,
     ) -> FederationResult<IncomingContractOffer>;
-    async fn fetch_consensus_clock_time(&self) -> FederationResult<SystemTime>;
+    async fn fetch_consensus_clock_time(&self) -> FederationResult<u64>;
```
Not sure if that is an improvement, more specific types are typically better, what was the reason here?

elsirion commented on modules/fedimint-ln-server/src/lib.rs:
```diff
@@ -908,23 +908,16 @@ impl Lightning {
-            _ => proposals[proposals.len() / 2],
-        };
-        median_proposal
+
+        *proposals
+            .get(proposals.len() / 2)
+            .expect("We checked before that proposals aren't empty")
```
Solves the previous comment about floats.

elsirion reviewed (commented):
I'm sorry about the confusion, but enforcement of expiry is most needed on the recipient side, that's why the issue and the [linked comment in there](https://github.com/fedimint/fedimint/pull/727#discussion_r989130513) mention *offers*, which belong to the *incoming* LN payment flow. The consensus stuff looks mostly good though, except for the one bug I wrote in the wallet and you only copied.

We could keep the outgoing timeout validation to avoid refunds by the gateway, although we wouldn't need the consensus time for that but could use the local time of the client.

justinmoon commented:
Are you planning to continue working on this @wilfredallyn

wilfredallyn commented:
Yes, I am going to revise this and finish it

wilfredallyn commented:
@justinmoon, I'm closing this after discussing with @joschisan. The increased complexity with a new consensus item (clock time) may not be worth it for this use case (enforcing invoice expiry). If this issue gets picked up in the future, these things need to be fixed:

- enforce expiry on incoming payment (was implemented on outgoing)
- update code to work with aleph bft
